### PR TITLE
NPY: Add check that buffer is long enough

### PR DESCRIPTION
**Pull request by kylebarron**
### Change list

- Add check that buffer is long enough when parsing NPY
- Improved return type

### CODE REVIEW

File: modules/textures/src/lib/parsers/parse-npy.ts [hunk 1]
```diff
@@ -1,6 +1,17 @@
 // import type {TextureLevel} from '@loaders.gl/schema';
 
 type NumpyHeader = {descr: string; shape: number[]};
+type ArrayTypes =
```
Comment (ibgreen) [inline modules/textures/src/lib/parsers/parse-npy.ts:4]: I suspect this type is availble in loader-utils or similar?

File: modules/textures/src/lib/parsers/parse-npy.ts [hunk 2]
```diff
@@ -48,6 +59,13 @@ export function parseNPY(arrayBuffer: ArrayBuffer, options?: unknown) {
   const nArrayElements = header.shape?.reduce((a: number, b: number): number => a * b);
   const arrayByteLength = nArrayElements * ArrayType.BYTES_PER_ELEMENT;
 
+  if (arrayBuffer.byteLength < headerEndOffset + arrayByteLength) {
+    // eslint-disable-next-line no-console, no-undef
+    console.warn(
+      `Buffer not long enough for slice: ${headerEndOffset}-${headerEndOffset + arrayByteLength}`
+    );
+    return null;
```
Comment (ibgreen) [inline modules/textures/src/lib/parsers/parse-npy.ts:67]: Is this better than throwing?

File: modules/textures/src/lib/parsers/parse-npy.ts [hunk 3]
```diff
@@ -48,6 +59,13 @@ export function parseNPY(arrayBuffer: ArrayBuffer, options?: unknown) {
   const nArrayElements = header.shape?.reduce((a: number, b: number): number => a * b);
   const arrayByteLength = nArrayElements * ArrayType.BYTES_PER_ELEMENT;
 
+  if (arrayBuffer.byteLength < headerEndOffset + arrayByteLength) {
+    // eslint-disable-next-line no-console, no-undef
+    console.warn(
+      `Buffer not long enough for slice: ${headerEndOffset}-${headerEndOffset + arrayByteLength}`
```
Comment (ibgreen) [inline modules/textures/src/lib/parsers/parse-npy.ts:65]: That is a very long and very technical message to add to the bundle.

**Review by ibgreen — APPROVED** [review]

File: modules/textures/src/lib/parsers/parse-npy.ts (hunk 1 above)
Comment (kylebarron) [reply modules/textures/src/lib/parsers/parse-npy.ts:4]: The problem is that here I want to refer to the _constructor_ of various typed arrays. TS doesn't seem to allow `typeof` on a union of types:

```ts
import type {TypedArray} from '@loaders.gl/loader-utils/src/types';
type ConstructorType = typeof TypedArray;
// Error: 'TypedArray' only refers to a type, but is being used as a value here.ts(2693)
```

File: modules/textures/src/lib/parsers/parse-npy.ts (hunk 2 above)
Comment (kylebarron) [reply modules/textures/src/lib/parsers/parse-npy.ts:67]: Previously we logged a warning and returned null, but we can change that

File: modules/textures/src/lib/parsers/parse-npy.ts (hunk 1 above)
Comment (kylebarron) [reply modules/textures/src/lib/parsers/parse-npy.ts:4]: There is a non-exported `TypedArrayConstructor` in gltf
https://github.com/visgl/loaders.gl/blob/fa3d2d0e464676b491f8581871058f9d3b8c79ec/modules/gltf/src/lib/gltf-utils/gltf-utils.ts#L5-L15

I'll rename this type here to be consistent

**Pull request merged**